### PR TITLE
Fix instruction counter to allow multiple register/memory accesses per instruction call

### PR DESCRIPTION
**Pull request by querolita**
This PR is meant to fix the subtraction underflows that appeared in unit tests when the same instruction was accessing the same register multiple times. This PR solves that issue by artificially increasing the instruction counter every time that register access takes place (so that the elapsed time is always a positive value). 

It compliments https://github.com/o1-labs/proof-systems/pull/2310. 

I tested it in a local branch with `LoadWordLeft` which used to fail for this reason (https://github.com/o1-labs/proof-systems/actions/runs/9447808562/job/26020430671?pr=2299), and now with the same seed the test `RUSTFLAGS="-Coverflow-checks=y -Cdebug-assertions=y" cargo test --package o1vm --lib --all-features --release -- mips::tests::unit::itype::test_unit_lwl_instruction --exact --show-output` works.

I will be updating the unit tests accordingly now, meaning
https://github.com/o1-labs/proof-systems/pull/2299
https://github.com/o1-labs/proof-systems/pull/2293
https://github.com/o1-labs/proof-systems/pull/2305

to check that this solves the issue.

Nonetheless, we need to run the whole demo to make sure that this fix does not break other parts of the system.

### CODE REVIEW

**Comment by dannywillems** [comment]
As discussed some min ago, and suggested by @mrmr1993, we should get the number of times we read/write/access a memory cell/register, and has the instruction counter being a multiple of it. It eases the debugging when we want to stop at instruction N for instance. See https://github.com/o1-labs/proof-systems/pull/2311#issuecomment-2160979048 for a branch printing the number of times we acess a register/memory cell. My testing shows it is 12 atm.

**Review by dannywillems — CHANGES_REQUESTED** [review]
See  comment + discussions.

**Review by dannywillems — CHANGES_REQUESTED** [review]
The log is wrong:
```
step=380000000 pc=00003391f8 insn=008c870080 ips=1759259.26
```

ips and step

**Comment by querolita** [comment]
> The log is wrong:

@dannywillems will fix it, forgot to adapt the log

File: o1vm/src/mips/witness.rs [hunk 1]
```diff
@@ -1050,6 +1064,24 @@ impl<Fp: Field, PreImageOracle: PreImageOracleT> Env<Fp, PreImageOracle> {
...
+    /// NOTE: actually, in practice it will be less than that, as there is no
+    ///       single instruction that performs all of them.
+    ///
+    /// This means that the actual number of instructions executed will result
+    /// from dividing the instruction counter by MAX_ACC (floor).
+    ///
+    /// Then, in order to update the instruction counter, we need to add 1 to
+    /// the real instruction counter and multiply it by MAX_ACC to have a unique
+    /// representation of each step (which is helpful for debugging).
+    pub fn update_instruction_counter(&mut self) {
```
Comment (dannywillems) [inline o1vm/src/mips/witness.rs:1081]: I think we might need to add another constraint. We might be able to add intermediate constraints. Will postpone the discussion to the soundness proof step, and consider this a good thing to move on.

**Review by dannywillems — APPROVED** [review]

File: o1vm/src/mips/witness.rs (hunk 1 above)
Comment (querolita) [reply o1vm/src/mips/witness.rs:1081]: https://github.com/o1-labs/proof-systems/issues/2322

**Pull request merged**